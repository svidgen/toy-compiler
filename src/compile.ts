import process from 'process';
import { readFileSync } from 'fs';
import MonkeyScript from './monkeyscript-lang';

const ast = MonkeyScript.parse({
	code: readFileSync(process.argv[2], 'utf-8').trim()
});

console.dir(ast, { depth: 2 });
