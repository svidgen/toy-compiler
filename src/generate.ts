import { AST } from './parser';

/**
 * The output.
 */

const output = (ast: AST): string => {
    switch (ast.type) {
        case 'GAME':
            return null;
        case 'STATEMENT':
        case 'STATEMENT_BODY':
            return ast.children
                .map(c => outputAsyncJS(c))
                .filter(o => o).join(";")
            ;
        case 'ASSIGNMENT':
            return ast.children
                .filter(c => c.type !== 'WHITESPACE')
                .map(c => outputAsyncJS(c)).join(" ")
            ;
        case 'ASSIGNMENT_OPERATOR':
            return '=';
        case 'NAME':
            return ast.code;
        case 'VALUE':
            return 'await ' + ast.code;
        case 'EOS':
        case 'WHITESPACE':
        case 'NOISE':
            return '';
        default:
            throw new Error(`Unknown token "${ast.type}":\n${ast.code}`)
    }
};

/**
 * Execution.
 */
const grammar = new Sequence('PROGRAM', STATEMENT);
const syntaxTree = grammar.parse({code: `
    xyz <- 123 ;;;

    ;  abc <- xyz
`});
const js = syntaxTree ? outputAsyncJS(syntaxTree) : '';

console.clear();
console.log({
    syntaxTree,
    js
});
