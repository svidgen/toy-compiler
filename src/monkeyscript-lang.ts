import {
	TreePattern,
	AST,
	Sequence,
	Union,
	Recipe,
	Token
} from './parser';


/**
 * The grammar.
 */
const EOS = new Token("EOS", /\s*[\n|;]+\s*/);
const ASSIGNMENT_OPERATOR = new Token("ASSIGNMENT_OPERATOR", /^=/);
const WHITESPACE = new Token("WHITESPACE", /^\s+/, true);
const NAME = new Token("NAME", /^[a-zA-Z_][a-zA-Z0-9_]*/);
const OPERATOR = new Token("OPERATOR", /^[+*]/);
const DEF_SEPARATOR = new Token("DEF_SEPARATOR", /^:/);
const INTEGER = new Token("INTEGER", /^[0-9]+/);
const FLOAT = new Token("FLOAT", /^[0-9]*\.[0-9]+/);
const NUMBER = new Union("NUMBER", [ INTEGER, FLOAT ]);
const VALUE = new Union("VALUE", [ NAME, NUMBER ]);

const EXPRESSION = new Union(
	"EXPRESSION",
	[ "OPERATION", NAME, VALUE ]
);

const OPERAND = new Union(
	"OPERAND",
	[ NAME, VALUE ]
);

const OPERATION = new Recipe(
	"OPERATION",
	[ OPERAND, WHITESPACE, OPERATOR, WHITESPACE, OPERAND ]
);

const ASSIGNMENT = new Recipe(
    "ASSIGNMENT",
    [ NAME, WHITESPACE, ASSIGNMENT_OPERATOR, WHITESPACE, EXPRESSION ]
);

const OP_DEFINITION = new Recipe(
	"OP_DEFINITION",
	[ 
		new Token("OP_KEYWORD", /^operation/i),
		DEF_SEPARATOR,
		WHITESPACE,
		ASSIGNMENT
	]
);

const ITEMS_DEFINITION = new Recipe(
	"ITEMS_DEFINITION",
	[
		new Token("OP_KEYWORD", /^starting items/i),
		DEF_SEPARATOR,
		WHITESPACE,
		"VALUE_LIST"
	]
);

const LIST_ITEM_SEPARATOR = new Token("LIST_ITEM_SEPARATOR", /^\s*,\s*/, true);

const VALUE_LIST = new Sequence(
	"VALUE_LIST",
	VALUE,
	LIST_ITEM_SEPARATOR
);

const STATEMENT = new Recipe("STATEMENT", [
    new Union("NOISE", [WHITESPACE, EOS]),
    new Union("STATEMENT_BODY", [
        ASSIGNMENT
    ]),
    EOS
]);

const TEST = new Recipe(
	"TEST",
	[
		new Token("TEST_BOILERPLATE", /^Test: divisible by/),
		WHITESPACE,
		NUMBER,
		WHITESPACE,
		new Token("TRUE_OPTION_BOILER", /^If true: throw to monkey/),
		WHITESPACE,
		NUMBER,
		WHITESPACE,
		new Token("FALSE_OPTION_BOILER", /^If false: throw to monkey/),
		WHITESPACE,
		NUMBER
	]
)

const MONKEY_KW = new Token("MONKEY_KW", /^monkey/i);
const MONKEY_ID = new Union("MONKEY_ID", [ INTEGER ]);
const MONKEY = new Recipe(
	"MONKEY",
	[
		MONKEY_KW, WHITESPACE, MONKEY_ID, DEF_SEPARATOR,
			WHITESPACE, ITEMS_DEFINITION,	
			WHITESPACE, OP_DEFINITION,	
			WHITESPACE, TEST,
	]
);

const GAME = new Sequence("GAME", MONKEY, WHITESPACE, false);

export default GAME;
