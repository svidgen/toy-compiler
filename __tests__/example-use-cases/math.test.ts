import { token, union, sequence, recipe } from '../../src/parser';

describe('basic math use-case example', () => {

    // Do not mistake this for a correct or good way to actually model math.
    // It's just an example.

    const OPTIONAL_WHITESPACE = token('WHITESPACE', /\s+/, true);
    const NUMBER = token('NUMBER', /\d+(\.\d+)?/);
    const ADDITION = token('ADDITION', '+');
    const SUBTRACTION = token('SUBTRACTION', '-');
    const MULTIPLICATION = token('MULTIPLICATION', '*');
    const DIVISION = token('DIVISION', '/');
    const OPERATOR = union('OPERATION', [ADDITION, SUBTRACTION, MULTIPLICATION, DIVISION]);
    const LEFT_PAREN = token('LEFT_PAREN', '(');
    const RIGHT_PAREN = token('RIGHT_PAREN', ')');
    const SUB_EXPRESSION = recipe('SUB_EXPRESSION', [
        LEFT_PAREN,
        OPTIONAL_WHITESPACE,
        'EXPRESSION',
        OPTIONAL_WHITESPACE,
        RIGHT_PAREN
    ])
    const TERM_MEMBER = union('TERM', [NUMBER, SUB_EXPRESSION])
    const TERM = recipe('TERM', [OPTIONAL_WHITESPACE, TERM_MEMBER, OPTIONAL_WHITESPACE])
    const EXPRESSION = sequence('EXPRESSION', TERM, OPERATOR);

    const expressions = [
        '1 + 2',
        '2 * 3',
        '2 * 1 + 3',
        '123 - (12 + 123) * 456',
        '(123 + 456) * (678 - 987)',
        '1+2',
        '2*3',
        '2*1+3',
        '123-(12+123)*456',
        '(123+456)*(678-987)',
    ] as const;

    for (const code of expressions) {
        it(`can parse ${code}`, () => {
            const ast = EXPRESSION.parse({code});
            expect(ast).toMatchSnapshot();
        })
    }
})