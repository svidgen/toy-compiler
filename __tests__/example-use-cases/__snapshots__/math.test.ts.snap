// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic math use-case example can parse (123 + 456) * (678 - 987) 1`] = `
AST {
  "children": [
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [
                AST {
                  "children": [],
                  "code": "(",
                  "end": 1,
                  "start": 0,
                  "type": "LEFT_PAREN",
                },
                AST {
                  "children": [
                    AST {
                      "children": [
                        AST {
                          "children": [
                            AST {
                              "children": [],
                              "code": "123",
                              "end": 4,
                              "start": 1,
                              "type": "NUMBER",
                            },
                          ],
                          "code": "123",
                          "end": 4,
                          "start": 1,
                          "type": "TERM",
                        },
                        AST {
                          "children": [],
                          "code": " ",
                          "end": 5,
                          "start": 4,
                          "type": "WHITESPACE",
                        },
                      ],
                      "code": "123 ",
                      "end": 5,
                      "start": 1,
                      "type": "TERM",
                    },
                    AST {
                      "children": [
                        AST {
                          "children": [],
                          "code": "+",
                          "end": 6,
                          "start": 5,
                          "type": "ADDITION",
                        },
                      ],
                      "code": "+",
                      "end": 6,
                      "start": 5,
                      "type": "OPERATION",
                    },
                    AST {
                      "children": [
                        AST {
                          "children": [],
                          "code": " ",
                          "end": 7,
                          "start": 6,
                          "type": "WHITESPACE",
                        },
                        AST {
                          "children": [
                            AST {
                              "children": [],
                              "code": "456",
                              "end": 10,
                              "start": 7,
                              "type": "NUMBER",
                            },
                          ],
                          "code": "456",
                          "end": 10,
                          "start": 7,
                          "type": "TERM",
                        },
                      ],
                      "code": " 456",
                      "end": 10,
                      "start": 6,
                      "type": "TERM",
                    },
                  ],
                  "code": "123 + 456",
                  "end": 10,
                  "start": 1,
                  "type": "EXPRESSION",
                },
                AST {
                  "children": [],
                  "code": ")",
                  "end": 11,
                  "start": 10,
                  "type": "RIGHT_PAREN",
                },
              ],
              "code": "(123 + 456)",
              "end": 11,
              "start": 0,
              "type": "SUB_EXPRESSION",
            },
          ],
          "code": "(123 + 456)",
          "end": 11,
          "start": 0,
          "type": "TERM",
        },
        AST {
          "children": [],
          "code": " ",
          "end": 12,
          "start": 11,
          "type": "WHITESPACE",
        },
      ],
      "code": "(123 + 456) ",
      "end": 12,
      "start": 0,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "*",
          "end": 13,
          "start": 12,
          "type": "MULTIPLICATION",
        },
      ],
      "code": "*",
      "end": 13,
      "start": 12,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": " ",
          "end": 14,
          "start": 13,
          "type": "WHITESPACE",
        },
        AST {
          "children": [
            AST {
              "children": [
                AST {
                  "children": [],
                  "code": "(",
                  "end": 15,
                  "start": 14,
                  "type": "LEFT_PAREN",
                },
                AST {
                  "children": [
                    AST {
                      "children": [
                        AST {
                          "children": [
                            AST {
                              "children": [],
                              "code": "678",
                              "end": 18,
                              "start": 15,
                              "type": "NUMBER",
                            },
                          ],
                          "code": "678",
                          "end": 18,
                          "start": 15,
                          "type": "TERM",
                        },
                        AST {
                          "children": [],
                          "code": " ",
                          "end": 19,
                          "start": 18,
                          "type": "WHITESPACE",
                        },
                      ],
                      "code": "678 ",
                      "end": 19,
                      "start": 15,
                      "type": "TERM",
                    },
                    AST {
                      "children": [
                        AST {
                          "children": [],
                          "code": "-",
                          "end": 20,
                          "start": 19,
                          "type": "SUBTRACTION",
                        },
                      ],
                      "code": "-",
                      "end": 20,
                      "start": 19,
                      "type": "OPERATION",
                    },
                    AST {
                      "children": [
                        AST {
                          "children": [],
                          "code": " ",
                          "end": 21,
                          "start": 20,
                          "type": "WHITESPACE",
                        },
                        AST {
                          "children": [
                            AST {
                              "children": [],
                              "code": "987",
                              "end": 24,
                              "start": 21,
                              "type": "NUMBER",
                            },
                          ],
                          "code": "987",
                          "end": 24,
                          "start": 21,
                          "type": "TERM",
                        },
                      ],
                      "code": " 987",
                      "end": 24,
                      "start": 20,
                      "type": "TERM",
                    },
                  ],
                  "code": "678 - 987",
                  "end": 24,
                  "start": 15,
                  "type": "EXPRESSION",
                },
                AST {
                  "children": [],
                  "code": ")",
                  "end": 25,
                  "start": 24,
                  "type": "RIGHT_PAREN",
                },
              ],
              "code": "(678 - 987)",
              "end": 25,
              "start": 14,
              "type": "SUB_EXPRESSION",
            },
          ],
          "code": "(678 - 987)",
          "end": 25,
          "start": 14,
          "type": "TERM",
        },
      ],
      "code": " (678 - 987)",
      "end": 25,
      "start": 13,
      "type": "TERM",
    },
  ],
  "code": "(123 + 456) * (678 - 987)",
  "end": 25,
  "start": 0,
  "type": "EXPRESSION",
}
`;

exports[`basic math use-case example can parse (123+456)*(678-987) 1`] = `
AST {
  "children": [
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [
                AST {
                  "children": [],
                  "code": "(",
                  "end": 1,
                  "start": 0,
                  "type": "LEFT_PAREN",
                },
                AST {
                  "children": [
                    AST {
                      "children": [
                        AST {
                          "children": [
                            AST {
                              "children": [],
                              "code": "123",
                              "end": 4,
                              "start": 1,
                              "type": "NUMBER",
                            },
                          ],
                          "code": "123",
                          "end": 4,
                          "start": 1,
                          "type": "TERM",
                        },
                      ],
                      "code": "123",
                      "end": 4,
                      "start": 1,
                      "type": "TERM",
                    },
                    AST {
                      "children": [
                        AST {
                          "children": [],
                          "code": "+",
                          "end": 5,
                          "start": 4,
                          "type": "ADDITION",
                        },
                      ],
                      "code": "+",
                      "end": 5,
                      "start": 4,
                      "type": "OPERATION",
                    },
                    AST {
                      "children": [
                        AST {
                          "children": [
                            AST {
                              "children": [],
                              "code": "456",
                              "end": 8,
                              "start": 5,
                              "type": "NUMBER",
                            },
                          ],
                          "code": "456",
                          "end": 8,
                          "start": 5,
                          "type": "TERM",
                        },
                      ],
                      "code": "456",
                      "end": 8,
                      "start": 5,
                      "type": "TERM",
                    },
                  ],
                  "code": "123+456",
                  "end": 8,
                  "start": 1,
                  "type": "EXPRESSION",
                },
                AST {
                  "children": [],
                  "code": ")",
                  "end": 9,
                  "start": 8,
                  "type": "RIGHT_PAREN",
                },
              ],
              "code": "(123+456)",
              "end": 9,
              "start": 0,
              "type": "SUB_EXPRESSION",
            },
          ],
          "code": "(123+456)",
          "end": 9,
          "start": 0,
          "type": "TERM",
        },
      ],
      "code": "(123+456)",
      "end": 9,
      "start": 0,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "*",
          "end": 10,
          "start": 9,
          "type": "MULTIPLICATION",
        },
      ],
      "code": "*",
      "end": 10,
      "start": 9,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [
                AST {
                  "children": [],
                  "code": "(",
                  "end": 11,
                  "start": 10,
                  "type": "LEFT_PAREN",
                },
                AST {
                  "children": [
                    AST {
                      "children": [
                        AST {
                          "children": [
                            AST {
                              "children": [],
                              "code": "678",
                              "end": 14,
                              "start": 11,
                              "type": "NUMBER",
                            },
                          ],
                          "code": "678",
                          "end": 14,
                          "start": 11,
                          "type": "TERM",
                        },
                      ],
                      "code": "678",
                      "end": 14,
                      "start": 11,
                      "type": "TERM",
                    },
                    AST {
                      "children": [
                        AST {
                          "children": [],
                          "code": "-",
                          "end": 15,
                          "start": 14,
                          "type": "SUBTRACTION",
                        },
                      ],
                      "code": "-",
                      "end": 15,
                      "start": 14,
                      "type": "OPERATION",
                    },
                    AST {
                      "children": [
                        AST {
                          "children": [
                            AST {
                              "children": [],
                              "code": "987",
                              "end": 18,
                              "start": 15,
                              "type": "NUMBER",
                            },
                          ],
                          "code": "987",
                          "end": 18,
                          "start": 15,
                          "type": "TERM",
                        },
                      ],
                      "code": "987",
                      "end": 18,
                      "start": 15,
                      "type": "TERM",
                    },
                  ],
                  "code": "678-987",
                  "end": 18,
                  "start": 11,
                  "type": "EXPRESSION",
                },
                AST {
                  "children": [],
                  "code": ")",
                  "end": 19,
                  "start": 18,
                  "type": "RIGHT_PAREN",
                },
              ],
              "code": "(678-987)",
              "end": 19,
              "start": 10,
              "type": "SUB_EXPRESSION",
            },
          ],
          "code": "(678-987)",
          "end": 19,
          "start": 10,
          "type": "TERM",
        },
      ],
      "code": "(678-987)",
      "end": 19,
      "start": 10,
      "type": "TERM",
    },
  ],
  "code": "(123+456)*(678-987)",
  "end": 19,
  "start": 0,
  "type": "EXPRESSION",
}
`;

exports[`basic math use-case example can parse 1 + 2 1`] = `
AST {
  "children": [
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [],
              "code": "1",
              "end": 1,
              "start": 0,
              "type": "NUMBER",
            },
          ],
          "code": "1",
          "end": 1,
          "start": 0,
          "type": "TERM",
        },
        AST {
          "children": [],
          "code": " ",
          "end": 2,
          "start": 1,
          "type": "WHITESPACE",
        },
      ],
      "code": "1 ",
      "end": 2,
      "start": 0,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "+",
          "end": 3,
          "start": 2,
          "type": "ADDITION",
        },
      ],
      "code": "+",
      "end": 3,
      "start": 2,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": " ",
          "end": 4,
          "start": 3,
          "type": "WHITESPACE",
        },
        AST {
          "children": [
            AST {
              "children": [],
              "code": "2",
              "end": 5,
              "start": 4,
              "type": "NUMBER",
            },
          ],
          "code": "2",
          "end": 5,
          "start": 4,
          "type": "TERM",
        },
      ],
      "code": " 2",
      "end": 5,
      "start": 3,
      "type": "TERM",
    },
  ],
  "code": "1 + 2",
  "end": 5,
  "start": 0,
  "type": "EXPRESSION",
}
`;

exports[`basic math use-case example can parse 1+2 1`] = `
AST {
  "children": [
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [],
              "code": "1",
              "end": 1,
              "start": 0,
              "type": "NUMBER",
            },
          ],
          "code": "1",
          "end": 1,
          "start": 0,
          "type": "TERM",
        },
      ],
      "code": "1",
      "end": 1,
      "start": 0,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "+",
          "end": 2,
          "start": 1,
          "type": "ADDITION",
        },
      ],
      "code": "+",
      "end": 2,
      "start": 1,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [],
              "code": "2",
              "end": 3,
              "start": 2,
              "type": "NUMBER",
            },
          ],
          "code": "2",
          "end": 3,
          "start": 2,
          "type": "TERM",
        },
      ],
      "code": "2",
      "end": 3,
      "start": 2,
      "type": "TERM",
    },
  ],
  "code": "1+2",
  "end": 3,
  "start": 0,
  "type": "EXPRESSION",
}
`;

exports[`basic math use-case example can parse 2 * 1 + 3 1`] = `
AST {
  "children": [
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [],
              "code": "2",
              "end": 1,
              "start": 0,
              "type": "NUMBER",
            },
          ],
          "code": "2",
          "end": 1,
          "start": 0,
          "type": "TERM",
        },
        AST {
          "children": [],
          "code": " ",
          "end": 2,
          "start": 1,
          "type": "WHITESPACE",
        },
      ],
      "code": "2 ",
      "end": 2,
      "start": 0,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "*",
          "end": 3,
          "start": 2,
          "type": "MULTIPLICATION",
        },
      ],
      "code": "*",
      "end": 3,
      "start": 2,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": " ",
          "end": 4,
          "start": 3,
          "type": "WHITESPACE",
        },
        AST {
          "children": [
            AST {
              "children": [],
              "code": "1",
              "end": 5,
              "start": 4,
              "type": "NUMBER",
            },
          ],
          "code": "1",
          "end": 5,
          "start": 4,
          "type": "TERM",
        },
        AST {
          "children": [],
          "code": " ",
          "end": 6,
          "start": 5,
          "type": "WHITESPACE",
        },
      ],
      "code": " 1 ",
      "end": 6,
      "start": 3,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "+",
          "end": 7,
          "start": 6,
          "type": "ADDITION",
        },
      ],
      "code": "+",
      "end": 7,
      "start": 6,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": " ",
          "end": 8,
          "start": 7,
          "type": "WHITESPACE",
        },
        AST {
          "children": [
            AST {
              "children": [],
              "code": "3",
              "end": 9,
              "start": 8,
              "type": "NUMBER",
            },
          ],
          "code": "3",
          "end": 9,
          "start": 8,
          "type": "TERM",
        },
      ],
      "code": " 3",
      "end": 9,
      "start": 7,
      "type": "TERM",
    },
  ],
  "code": "2 * 1 + 3",
  "end": 9,
  "start": 0,
  "type": "EXPRESSION",
}
`;

exports[`basic math use-case example can parse 2 * 3 1`] = `
AST {
  "children": [
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [],
              "code": "2",
              "end": 1,
              "start": 0,
              "type": "NUMBER",
            },
          ],
          "code": "2",
          "end": 1,
          "start": 0,
          "type": "TERM",
        },
        AST {
          "children": [],
          "code": " ",
          "end": 2,
          "start": 1,
          "type": "WHITESPACE",
        },
      ],
      "code": "2 ",
      "end": 2,
      "start": 0,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "*",
          "end": 3,
          "start": 2,
          "type": "MULTIPLICATION",
        },
      ],
      "code": "*",
      "end": 3,
      "start": 2,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": " ",
          "end": 4,
          "start": 3,
          "type": "WHITESPACE",
        },
        AST {
          "children": [
            AST {
              "children": [],
              "code": "3",
              "end": 5,
              "start": 4,
              "type": "NUMBER",
            },
          ],
          "code": "3",
          "end": 5,
          "start": 4,
          "type": "TERM",
        },
      ],
      "code": " 3",
      "end": 5,
      "start": 3,
      "type": "TERM",
    },
  ],
  "code": "2 * 3",
  "end": 5,
  "start": 0,
  "type": "EXPRESSION",
}
`;

exports[`basic math use-case example can parse 2*1+3 1`] = `
AST {
  "children": [
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [],
              "code": "2",
              "end": 1,
              "start": 0,
              "type": "NUMBER",
            },
          ],
          "code": "2",
          "end": 1,
          "start": 0,
          "type": "TERM",
        },
      ],
      "code": "2",
      "end": 1,
      "start": 0,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "*",
          "end": 2,
          "start": 1,
          "type": "MULTIPLICATION",
        },
      ],
      "code": "*",
      "end": 2,
      "start": 1,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [],
              "code": "1",
              "end": 3,
              "start": 2,
              "type": "NUMBER",
            },
          ],
          "code": "1",
          "end": 3,
          "start": 2,
          "type": "TERM",
        },
      ],
      "code": "1",
      "end": 3,
      "start": 2,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "+",
          "end": 4,
          "start": 3,
          "type": "ADDITION",
        },
      ],
      "code": "+",
      "end": 4,
      "start": 3,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [],
              "code": "3",
              "end": 5,
              "start": 4,
              "type": "NUMBER",
            },
          ],
          "code": "3",
          "end": 5,
          "start": 4,
          "type": "TERM",
        },
      ],
      "code": "3",
      "end": 5,
      "start": 4,
      "type": "TERM",
    },
  ],
  "code": "2*1+3",
  "end": 5,
  "start": 0,
  "type": "EXPRESSION",
}
`;

exports[`basic math use-case example can parse 2*3 1`] = `
AST {
  "children": [
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [],
              "code": "2",
              "end": 1,
              "start": 0,
              "type": "NUMBER",
            },
          ],
          "code": "2",
          "end": 1,
          "start": 0,
          "type": "TERM",
        },
      ],
      "code": "2",
      "end": 1,
      "start": 0,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "*",
          "end": 2,
          "start": 1,
          "type": "MULTIPLICATION",
        },
      ],
      "code": "*",
      "end": 2,
      "start": 1,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [],
              "code": "3",
              "end": 3,
              "start": 2,
              "type": "NUMBER",
            },
          ],
          "code": "3",
          "end": 3,
          "start": 2,
          "type": "TERM",
        },
      ],
      "code": "3",
      "end": 3,
      "start": 2,
      "type": "TERM",
    },
  ],
  "code": "2*3",
  "end": 3,
  "start": 0,
  "type": "EXPRESSION",
}
`;

exports[`basic math use-case example can parse 123 - (12 + 123) * 456 1`] = `
AST {
  "children": [
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [],
              "code": "123",
              "end": 3,
              "start": 0,
              "type": "NUMBER",
            },
          ],
          "code": "123",
          "end": 3,
          "start": 0,
          "type": "TERM",
        },
        AST {
          "children": [],
          "code": " ",
          "end": 4,
          "start": 3,
          "type": "WHITESPACE",
        },
      ],
      "code": "123 ",
      "end": 4,
      "start": 0,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "-",
          "end": 5,
          "start": 4,
          "type": "SUBTRACTION",
        },
      ],
      "code": "-",
      "end": 5,
      "start": 4,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": " ",
          "end": 6,
          "start": 5,
          "type": "WHITESPACE",
        },
        AST {
          "children": [
            AST {
              "children": [
                AST {
                  "children": [],
                  "code": "(",
                  "end": 7,
                  "start": 6,
                  "type": "LEFT_PAREN",
                },
                AST {
                  "children": [
                    AST {
                      "children": [
                        AST {
                          "children": [
                            AST {
                              "children": [],
                              "code": "12",
                              "end": 9,
                              "start": 7,
                              "type": "NUMBER",
                            },
                          ],
                          "code": "12",
                          "end": 9,
                          "start": 7,
                          "type": "TERM",
                        },
                        AST {
                          "children": [],
                          "code": " ",
                          "end": 10,
                          "start": 9,
                          "type": "WHITESPACE",
                        },
                      ],
                      "code": "12 ",
                      "end": 10,
                      "start": 7,
                      "type": "TERM",
                    },
                    AST {
                      "children": [
                        AST {
                          "children": [],
                          "code": "+",
                          "end": 11,
                          "start": 10,
                          "type": "ADDITION",
                        },
                      ],
                      "code": "+",
                      "end": 11,
                      "start": 10,
                      "type": "OPERATION",
                    },
                    AST {
                      "children": [
                        AST {
                          "children": [],
                          "code": " ",
                          "end": 12,
                          "start": 11,
                          "type": "WHITESPACE",
                        },
                        AST {
                          "children": [
                            AST {
                              "children": [],
                              "code": "123",
                              "end": 15,
                              "start": 12,
                              "type": "NUMBER",
                            },
                          ],
                          "code": "123",
                          "end": 15,
                          "start": 12,
                          "type": "TERM",
                        },
                      ],
                      "code": " 123",
                      "end": 15,
                      "start": 11,
                      "type": "TERM",
                    },
                  ],
                  "code": "12 + 123",
                  "end": 15,
                  "start": 7,
                  "type": "EXPRESSION",
                },
                AST {
                  "children": [],
                  "code": ")",
                  "end": 16,
                  "start": 15,
                  "type": "RIGHT_PAREN",
                },
              ],
              "code": "(12 + 123)",
              "end": 16,
              "start": 6,
              "type": "SUB_EXPRESSION",
            },
          ],
          "code": "(12 + 123)",
          "end": 16,
          "start": 6,
          "type": "TERM",
        },
        AST {
          "children": [],
          "code": " ",
          "end": 17,
          "start": 16,
          "type": "WHITESPACE",
        },
      ],
      "code": " (12 + 123) ",
      "end": 17,
      "start": 5,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "*",
          "end": 18,
          "start": 17,
          "type": "MULTIPLICATION",
        },
      ],
      "code": "*",
      "end": 18,
      "start": 17,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": " ",
          "end": 19,
          "start": 18,
          "type": "WHITESPACE",
        },
        AST {
          "children": [
            AST {
              "children": [],
              "code": "456",
              "end": 22,
              "start": 19,
              "type": "NUMBER",
            },
          ],
          "code": "456",
          "end": 22,
          "start": 19,
          "type": "TERM",
        },
      ],
      "code": " 456",
      "end": 22,
      "start": 18,
      "type": "TERM",
    },
  ],
  "code": "123 - (12 + 123) * 456",
  "end": 22,
  "start": 0,
  "type": "EXPRESSION",
}
`;

exports[`basic math use-case example can parse 123-(12+123)*456 1`] = `
AST {
  "children": [
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [],
              "code": "123",
              "end": 3,
              "start": 0,
              "type": "NUMBER",
            },
          ],
          "code": "123",
          "end": 3,
          "start": 0,
          "type": "TERM",
        },
      ],
      "code": "123",
      "end": 3,
      "start": 0,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "-",
          "end": 4,
          "start": 3,
          "type": "SUBTRACTION",
        },
      ],
      "code": "-",
      "end": 4,
      "start": 3,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [
                AST {
                  "children": [],
                  "code": "(",
                  "end": 5,
                  "start": 4,
                  "type": "LEFT_PAREN",
                },
                AST {
                  "children": [
                    AST {
                      "children": [
                        AST {
                          "children": [
                            AST {
                              "children": [],
                              "code": "12",
                              "end": 7,
                              "start": 5,
                              "type": "NUMBER",
                            },
                          ],
                          "code": "12",
                          "end": 7,
                          "start": 5,
                          "type": "TERM",
                        },
                      ],
                      "code": "12",
                      "end": 7,
                      "start": 5,
                      "type": "TERM",
                    },
                    AST {
                      "children": [
                        AST {
                          "children": [],
                          "code": "+",
                          "end": 8,
                          "start": 7,
                          "type": "ADDITION",
                        },
                      ],
                      "code": "+",
                      "end": 8,
                      "start": 7,
                      "type": "OPERATION",
                    },
                    AST {
                      "children": [
                        AST {
                          "children": [
                            AST {
                              "children": [],
                              "code": "123",
                              "end": 11,
                              "start": 8,
                              "type": "NUMBER",
                            },
                          ],
                          "code": "123",
                          "end": 11,
                          "start": 8,
                          "type": "TERM",
                        },
                      ],
                      "code": "123",
                      "end": 11,
                      "start": 8,
                      "type": "TERM",
                    },
                  ],
                  "code": "12+123",
                  "end": 11,
                  "start": 5,
                  "type": "EXPRESSION",
                },
                AST {
                  "children": [],
                  "code": ")",
                  "end": 12,
                  "start": 11,
                  "type": "RIGHT_PAREN",
                },
              ],
              "code": "(12+123)",
              "end": 12,
              "start": 4,
              "type": "SUB_EXPRESSION",
            },
          ],
          "code": "(12+123)",
          "end": 12,
          "start": 4,
          "type": "TERM",
        },
      ],
      "code": "(12+123)",
      "end": 12,
      "start": 4,
      "type": "TERM",
    },
    AST {
      "children": [
        AST {
          "children": [],
          "code": "*",
          "end": 13,
          "start": 12,
          "type": "MULTIPLICATION",
        },
      ],
      "code": "*",
      "end": 13,
      "start": 12,
      "type": "OPERATION",
    },
    AST {
      "children": [
        AST {
          "children": [
            AST {
              "children": [],
              "code": "456",
              "end": 16,
              "start": 13,
              "type": "NUMBER",
            },
          ],
          "code": "456",
          "end": 16,
          "start": 13,
          "type": "TERM",
        },
      ],
      "code": "456",
      "end": 16,
      "start": 13,
      "type": "TERM",
    },
  ],
  "code": "123-(12+123)*456",
  "end": 16,
  "start": 0,
  "type": "EXPRESSION",
}
`;
